acceptance/lib/helper.rb

```diff
@@ -413,7 +415,177 @@ def append_match_request(args)
+      http.puppet-v3-report-.*.-requests
+      http.puppet-v3-static_file_content-.*.-requests
+      http.total-requests
+      jruby.borrow-timer
+      jruby.free-jrubies-histo
+      jruby.lock-held-timer
+      jruby.lock-wait-timer
+      jruby.requested-jrubies-histo
+      jruby.wait-timer
+    )
+  end
+
+  def single_field_metrics
+    %w(
+      http.active-requests.count
+      http.puppet-v3-catalog-.*.-percentage
+      http.puppet-v3-environment-.*.-percentage
+      http.puppet-v3-environment_classes-.*.-percentage
+      http.puppet-v3-environments-percentage
+      http.puppet-v3-file_bucket_file-.*.-percentage
+      http.puppet-v3-file_content-.*.-percentage
+      http.puppet-v3-file_metadata-.*.-percentage
+      http.puppet-v3-file_metadatas-.*.-percentage
+      http.puppet-v3-node-.*.-percentage
+      http.puppet-v3-report-.*.-percentage
+      http.puppet-v3-resource_type-.*.-percentage
+      http.puppet-v3-resource_types-.*.-percentage
+      http.puppet-v3-static_file_content-.*.-percentage
+      http.puppet-v3-status-.*.-percentage
+      num-cpus
+      uptime
+      jruby.borrow-count.count
+      jruby.borrow-retry-count.count
+      jruby.borrow-timeout-count.count
+      jruby.request-count.count
+      jruby.return-count.count
+      jruby.num-free-jrubies
+      jruby.num-jrubies
+    )
+  end
 
+  # returns an array of default metrics by building the standard field based
+  # metrics, then appending the single field metrics
+  def build_default_metrics
+    default_metrics = Array.new
+    standard_metrics.each do |sm|
+      metrics_fields.each do |f|
+        default_metrics << "#{sm}.#{f}"
+      end
+    end
+    default_metrics = default_metrics + single_field_metrics
+    return default_metrics
+  end
+
+  # Strips the nils recorded by the graphite system that occur when the
+  # graphite system is recording on a faster cadence than the puppetserver
+  # system
+  def strip_nils(json)
+    json.each do |json_member|
+      json_member['datapoints'].each_index do |i|
+        while ( json_member['datapoints'][i] != nil &&
+            json_member['datapoints'][i][0] == nil ) do
+          json_member['datapoints'].delete_at(i)
+        end
+      end
+    end
+  end
+
+  def metrics_target(master_hostname, metric_name)
+    return_string = "puppetlabs.#{master_hostname}.#{metric_name}"
+  end
+
+  def build_metrics_url(master, graphite_host, metrics, minutes)
+    url = "http://#{graphite_host}/render?"
+    url += "from=-#{minutes}minutes"
+    if metrics.is_a?(String)
+      url += '&target=' << metrics_target(master, metrics)
+    elsif metrics.is_a?(Array)
+      metrics.each do |m|
+        url += '&target=' << metrics_target(master, m)
+      end
+    end
+    url += '&format=json'
+  end
+
+  # returns a hash of metrics results
+  def query_metrics(master_hostname, graphite_hostname, metrics, minutes=10)
+    sleep_between_queries = 5
+    seconds_remaining = minutes * 60
```

Comment:
Ah, I think the bottleneck was coming from Graphite, not Puppet Server.  If I configure the `MAX_CACHE_SIZE` setting in `/opt/graphite/conf/carbon.conf` to "inf" instead of the default, "50", the whole metric set is created in under 2 seconds.  Will see how best to work that into the Graphite configuration for this test.